### PR TITLE
Adding new rule: require-ember-lifeline

### PR DESCRIPTION
**Pull request by scalvert**
This rule identifies the use of `Ember.run.*` and indicates the requirement to use `runTask`, `debounceTask`, or `throttleTask` from ember-lifeline in favor of the run variant.

### CODE REVIEW

File: guides/rules/require-ember-lifeline.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
+# Require Ember Lifeline
+
+[Ember lifeline](https://github.com/rwjblue/ember-lifeline)
+
+Ember applications have long life-cycles. A user may navigate to several pages and use many different features before they leave the application. This makes JavaScript and Ember development unlike Rails development, where the lifecycle of a request is short and the environment disposed of after each request. It makes Ember development much more like iOS or video game development than traditional server-side web development.
+
+It is good to note that this isn't something inherent to just Ember. Any single-page app framework or solution (Angular, React, Vue, Backbone...) must deal this lifecycles of objects, and specifically with how async tasks can be bounded by a lifecycle.
```
Comment (trentmwillis) [inline guides/rules/require-ember-lifeline.md]: Nit: "must deal this lifecycles" -> "must deal with the lifecycles"

File: lib/rules/require-ember-lifeline.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    return {
+      ObjectPattern(node) {
+        if (emberImportBinding) {
+          LATER_PATTERNS = LATER_PATTERNS.concat(collectObjectPatternBindings(node, {
+            [emberImportBinding]: ['run']
+          }));
+        }
+      },
+
+      ImportDefaultSpecifier(node) {
```
Comment (trentmwillis) [inline lib/rules/require-ember-lifeline.js]: Nit: Can we move this above `ObjectPattern`? Since it is expected to run prior to that function.

File: lib/rules/require-ember-lifeline.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  meta: {
+    message: MESSAGE
+  },
+  create(context) {
+    let emberImportBinding;
+
+    return {
+      ObjectPattern(node) {
+        if (emberImportBinding) {
+          LATER_PATTERNS = LATER_PATTERNS.concat(collectObjectPatternBindings(node, {
```
Comment (trentmwillis) [inline lib/rules/require-ember-lifeline.js]: If I understand this correctly, this code catches if `run.later` was bound to something else, correct? If so, can we add a test for this scenario?

File: guides/rules/require-ember-lifeline.md (hunk 1 above)
Comment (scalvert) [inline guides/rules/require-ember-lifeline.md]: 👍

File: lib/rules/require-ember-lifeline.js (hunk 3 above)
Comment (scalvert) [inline lib/rules/require-ember-lifeline.js]: Yep, on it!

File: lib/rules/require-ember-lifeline.js (hunk 2 above)
Comment (scalvert) [inline lib/rules/require-ember-lifeline.js]: Yessir!

File: lib/rules/require-ember-lifeline.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
+/**
+ * @fileOverview Require the use of ember-lifeline's runTask over using Ember.run.later.
+ * @author Steve Calvert
```
Comment (trentmwillis) [inline lib/rules/require-ember-lifeline.js]: After spending a bit of time thinking about it, I'd like for us to stop placing `@author` annotations in these files. They feel territorial and might off put newer contributors.

If we want to acknowledge authors/contributors, we should go the route of placing an `AUTHORS.txt` file in the root of the repo which can be generated from the git history ([example from QUnit](https://github.com/qunitjs/qunit/blob/master/AUTHORS.txt)).

Comment (lynchbomb) [inline lib/rules/require-ember-lifeline.js]: Totally agree. AUTHORS.txt works for me.

File: lib/rules/require-ember-lifeline.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+const { getEmberImportBinding } = require('../utils/imports');
+const { collectObjectPatternBindings } = require('../utils/destructed-binding');
+
+let LATER_PATTERNS = ['Ember.run.later', 'run.later'];
+
+const MESSAGE = 'Please use runTask from ember-lifeline in favor of Ember.run.later.';
+
+module.exports = {
+  docs: {
+    description: 'Please use runTask from ember-lifeline in favor of Ember.run.later.',
```
Comment (lynchbomb) [inline lib/rules/require-ember-lifeline.js]: Can you leverage the `MESSAGE` variable rather than duplicate this message?

File: tests/lib/rules/require-ember-lifeline.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    {
+      code: `
+        export default Ember.Component({
+          init() {
+            this.runTask(() => {
+              doSomeWork();
+            });
+          }
+        });`,
+      parserOptions: {
```
Comment (lynchbomb) [inline tests/lib/rules/require-ember-lifeline.js]: Can we move the `parserOptions` up into the `RuleTester()` and then remove the duplicated `parserOptions`?

example: 
```js
const ruleTester = new RuleTester({
  parserOptions: { ... }
});
```

Comment (scalvert) [inline tests/lib/rules/require-ember-lifeline.js]: Fo sho.

**Comment by scalvert** [comment]
@trentmwillis updated to include both `debounce` and `throttle`. Can you take another look?

File: lib/rules/require-ember-lifeline.js (hunk 4 above)
Comment (scalvert) [inline lib/rules/require-ember-lifeline.js]: But how will the world know about my amazing feats of coding? 

Totally agree. I'll remove from all rules in a separate PR.

File: lib/rules/require-ember-lifeline.js (hunk 5 above)
Comment (scalvert) [inline lib/rules/require-ember-lifeline.js]: As you can see now, the message is dynamically generated based on the node being inspected. I've adjusted it now.

**Comment by scalvert** [comment]
Bump @trentmwillis

File: tests/lib/rules/require-ember-lifeline.js [hunk 7]
```diff
@@ -1,26 +1,31 @@
...
             this.runTask(() => {
               doSomeWork();
             });
           }
         });`,
-      parserOptions: {
-        ecmaVersion: 6,
-        sourceType: 'module'
-      }
+      parserOptions
```
Comment (lynchbomb) [inline tests/lib/rules/require-ember-lifeline.js]: Can we delete these redundant `parserOptions`?

File: tests/lib/rules/require-ember-lifeline.js [hunk 8]
```diff
@@ -1,26 +1,31 @@
 const rule = require('../../../lib/rules/require-ember-lifeline');
-const MESSAGE = rule.meta.message;
+const getMessage = rule.meta.message;
 const RuleTester = require('eslint').RuleTester;
 const ruleTester = new RuleTester();
```
Comment (lynchbomb) [inline tests/lib/rules/require-ember-lifeline.js]: Can we just pass the `parserOptions` object directly into the instantiation of `RuleTester()`. That way we can delete the redundant `parserOptions` from each assertion.

Comment (scalvert) [inline tests/lib/rules/require-ember-lifeline.js]: We *can*, but do we care? If we do, we should do this across all the test files. That should be done separately in a different PR. Thoughts?

File: tests/lib/rules/require-ember-lifeline.js (hunk 7 above)
Comment (scalvert) [inline tests/lib/rules/require-ember-lifeline.js]: Same as above.

File: tests/lib/rules/require-ember-lifeline.js (hunk 8 above)
Comment (lynchbomb) [inline tests/lib/rules/require-ember-lifeline.js]: IDK I think it's more than a nit. We should prob make this change to all the tests. Figure it's still early days within this repo and we can solidify the ideals.

Essentially we have lines of code within tests that are redundant. I'm all for a balance of succinct yet grokkable. Thoughts?

Comment (scalvert) [inline tests/lib/rules/require-ember-lifeline.js]: I'm happy to refactor this, just want it to be in a separate PR. I created #65 to address this.

File: lib/rules/require-ember-lifeline.js [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+  meta: {
+    message: getMessage
+  },
+  create(context) {
+    let emberImportBinding;
+    let disallowedCalls = mergeDisallowedCalls(DISALLOWED_OBJECTS);
+
+    return {
+      ImportDefaultSpecifier(node) {
+        emberImportBinding = getEmberImportBinding(node);
```
Comment (rwjblue) [inline lib/rules/require-ember-lifeline.js]: We should also handle:

```js
import run from '@ember/runloop';
import { debounce } from '@ember/runloop';
import { throttle } from '@ember/runloop';
import { later } from '@ember/runloop';
import { next } from '@ember/runloop';
```

Could be done in a follow up PR though...

Comment (scalvert) [inline lib/rules/require-ember-lifeline.js]: Sounds good.

Comment (scalvert) [inline lib/rules/require-ember-lifeline.js]: Added #66

File: lib/rules/require-ember-lifeline.js (hunk 4 above)
Comment (scalvert) [inline lib/rules/require-ember-lifeline.js]: This is done.

File: lib/rules/require-ember-lifeline.js (hunk 3 above)
Comment (scalvert) [inline lib/rules/require-ember-lifeline.js]: Done

File: lib/rules/require-ember-lifeline.js (hunk 2 above)
Comment (scalvert) [inline lib/rules/require-ember-lifeline.js]: Done

File: tests/lib/rules/require-ember-lifeline.js (hunk 6 above)
Comment (scalvert) [inline tests/lib/rules/require-ember-lifeline.js]: Done

Comment (scalvert) [inline tests/lib/rules/require-ember-lifeline.js]: Done
